File: threeSum.js
var threeSum = function(nums) {
    let result = [];
    nums = nums.sort((a,b) => {
        return a-b
    })
    
    let dict = {};
    let solutionsDict = {}
    
    for (number of nums) {
        if (dict[number]) dict[number]++
        else dict[number] = 1;
    }

    
    for (let i = 0; i < nums.length; i++){
        // only test from negative to 0. because testing positive numbers will be redudant at this point
        if (nums[i] > 0) break;
        // skip duplicates, since we checked them already. 
        if (nums[i] === nums[i-1]) {
            dict[nums[i]]--;
            continue
        }

        dict[nums[i]]--;
        for (let j = i+1; j < nums.length; j++) {
            dict[nums[j]]--;
            let thirdNum = 0-nums[i]-nums[j];
            if (dict[thirdNum] >= 1) {
                let soln = [nums[i], nums[j], thirdNum];
             
                let stringSoln = JSON.stringify(soln);
                if (!solutionsDict[stringSoln]) {
                    result.push(soln);
                    solutionsDict[stringSoln] = true;
                }
               
            }
        }
        // add back the numbers we removed
        for (let k = i+1; k < nums.length; k++) {
            dict[nums[k]]++
        }
    
    }

    
    return result;   
};

// This code is still too slow!
// time complexity is O(n^2) for this code
// https://leetcode.com/submissions/detail/302874951/testcase/
// 3000 elements in the arr
// it needs to be able to handle this many cases

File: threeSum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        
        
        """
        
        result = []
        
        if len(nums) < 3:
            return []
        
        nums.sort()
        options = {}
        duplicates = {}
        
        for i in range(len(nums)):
            if nums[i] not in options:
                options[nums[i]] = [i]
            else:
                options[nums[i]].append(i)
        
        
        for i in range(len(nums)):
            if nums[i] > 0:
                break 
            elif nums[i] == 0 and str([0,0,0]) in duplicates:
                continue
            for j in range(i+1,len(nums)):
                if nums[i] == 0 and str([0,0,0]) in duplicates:
                    continue
                twoSum = -(nums[i] + nums[j])
                if twoSum in options:
                    arr = options[twoSum]
                    for k in range(len(arr)):
                        res = [nums[i],nums[j],nums[arr[k]]]
                        if arr[k] > j and str(res) not in duplicates:
                            result.append(res)
                            duplicates[str(res)] = True
        return result